### PR TITLE
[webapp-perf] Change anonymous inline function into a defined arrow function

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
- Change anonymous inline functions into a defined arrow functions.
- Make use of `data-attr` in storing information which serves as parameter of underlying function.

Note: This is second installment PR. Other PR submitted:
- https://github.com/mattermost/mattermost-webapp/pull/480

#### Ticket Link
[Please link the GitHub issue or Jira ticket this PR addresses.]

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: components/multiselect/multiselect.jsx [hunk 1]
```diff
@@ -238,8 +242,8 @@ export default class MultiSelect extends React.Component {
                             onChange={this.onChange}
                             value={this.props.values}
                             valueRenderer={this.props.valueRenderer}
-                            menuRenderer={() => null}
-                            arrowRenderer={() => null}
+                            menuRenderer={this.handleRender}
```
Comment (enahum) [inline components/multiselect/multiselect.jsx]: No need for a function here, actually you could even remove the props

File: components/navbar/navbar.jsx [hunk 2]
```diff
@@ -354,6 +354,10 @@ export default class Navbar extends React.Component {
         );
     }
 
+    hideHeaderOverlay = () => {
+        this.refs.headerOverlay.hide();
```
Comment (enahum) [inline components/navbar/navbar.jsx]: Make sure the ref exists

File: components/navbar/navbar.jsx [hunk 3]
```diff
@@ -791,6 +795,14 @@ export default class Navbar extends React.Component {
         return null;
     }
 
+    hideEditChannelHeaderModal = () => {
+        this.setState({showEditChannelHeaderModal: false});
+    }
+
+    showChannelInviteModalButton = () => {
+        this.refs.channelInviteModalButton.show();
```
Comment (enahum) [inline components/navbar/navbar.jsx]: Same here

File: components/new_channel_modal/new_channel_modal.jsx [hunk 4]
```diff
@@ -144,6 +144,11 @@ export default class NewChannelModal extends React.PureComponent {
         this.props.onDataChanged(newData);
     }
 
+    handleOnURLChange = (e) => {
+        e.preventDefault();
+        this.props.onChangeURLPressed();
```
Comment (enahum) [inline components/new_channel_modal/new_channel_modal.jsx]: Is the prop required? If not check that is defined

File: components/rhs_root_post.jsx [hunk 5]
```diff
@@ -211,6 +211,10 @@ export default class RhsRootPost extends React.Component {
         });
     }
 
+    getDotMenuRef = () => {
+        return this.refs.dotMenu;
```
Comment (enahum) [inline components/rhs_root_post.jsx]: Here too

File: components/multiselect/multiselect.jsx (hunk 1 above)
Comment (saturninoabril) [reply components/multiselect/multiselect.jsx]: I've tried to remove those but
- when `menuRenderer` is removed, the menu opened up as soon as the user access a component with MultiSelect (see screenshot)
screen shot 2017-12-19 at 1 16 44 am
- when `arrowRenderer`, an extra caret show up in the text box 
screen shot 2017-12-19 at 2 20 48 am

So I'll keep it that way.

File: components/rhs_root_post.jsx (hunk 5 above)
Comment (saturninoabril) [reply components/rhs_root_post.jsx]: Just add check at consuming component - `EmojiPickerOverlay`

**Comment by mkraft** [comment]
@saturninoabril Have you ensured that every component you're changing the prop to non-inline functions for fits one of the following:

a) Is a `PureComponent`, or
b) Is otherwise leveraging the prop with the new non-inline function in its `shouldComponentUpdate` method?

The reason I ask is that if neither of those two cases are met it won't be fully leveraging the performance benefit.

**Comment by saturninoabril** [comment]
@mkraft This PR is mainly an outcome of what was discussed (and I supposed was agreed upon) last performance meeting about inline function. A similar follow up discussion here - https://pre-release.mattermost.com/core/pl/s4yfjinzhby38887au35yn8yzo
You're absolutely right about the 2 points, especially for those that renders standard element (e.g. `<button onClick={this.handleClick}/>`).
This PR might be a premature optimization (without actually presenting metrics) but I guess the key take away/pattern is to avoid inline function on render method that creates new function every time a component renders; an inline function without referential identity. Not to mention but doing so makes the function becomes easier to read/understand.

**Comment by mkraft** [comment]
@saturninoabril I agree with what you're saying and maybe I should leave the `PureComponent` discussion as a separate discussion because certainly the changes in the PR have merit in their own right. It's just something for us to be aware of and I'm sure we'll address it in good time. Nice work 👍

**Comment by saturninoabril** [comment]
Rebase to latest master and fix merge conflicts

File: components/navbar/navbar.jsx (hunk 2 above)
Comment (enahum) [reply components/navbar/navbar.jsx]: what I mean here is that you should wrap this in an if statement like

```js
hideHeaderOverlay = () => {
   if (this.refs.headerOverlay) {
      this.refs.headerOverlay.hide();
   }
};
```

File: components/rhs_root_post.jsx (hunk 5 above)
Comment (enahum) [reply components/rhs_root_post.jsx]: I'm ok here

**Comment by enahum** [comment]
@saturninoabril the rest of the code looks good, but I don't feel very comfortable to merge it for 4.6 as is very late in the game and that PR looks sensible and it does not give an important performance boost, what do you think?

File: components/navbar/navbar.jsx (hunk 2 above)
Comment (saturninoabril) [reply components/navbar/navbar.jsx]: I totally missed this one. Should be good now.

**Pull request merged**